[[ SocketLocalPort ]] Added the ability to specify the local host and…

livecodemichael opened the pull request:
… port when opening a new socket.

Updated the parsing of open socket so that it now accepts the following form:
open socket [from [<local host>][:<local port>]] to <host>[:<port>]

Reworked the name to sockaddr parsing into:
- name to host and port
- host and port to sockaddr
Name to sockaddr now calls name to host and port and then host and port to
sockaddr.

The reworked name to sockaddr parsing is used to parse any local from details in
MCS_connect_socket: It checks to see if any local host or port has been passed.
If no local host has been passed, then we use, if set, the networkinterface
property. It then binds the socket to any found local host and port.

livecodemichael committed:
[[ SocketLocalPort ]] Added the ability to specify the local host and port when
opening a new socket.

Updated the parsing of open socket so that it now accepts the following form:
open socket [from [<local host>][:<local port>]] to <host>[:<port>]

Reworked the name to sockaddr parsing into:
- name to host and port
- host and port to sockaddr
Name to sockaddr now calls name to host and port and then host and port to
sockaddr.

The reworked name to sockaddr parsing is used to parse any local from details in
MCS_connect_socket: It checks to see if any local host or port has been passed.
If no local host has been passed, then we use, if set, the networkinterface
property. It then binds the socket to any found local host and port.

livecodemichael committed:
[[ SocketLocalPort ]] Updated open socket docs entry and added
release note.

peter-b commented on engine/src/cmds.h:
```diff
@@ -1530,6 +1530,7 @@ class MCOpen : public MCStatement
 	
 	// MM-2014-06-13: [[ Bug 12567 ]] Added new "open socket <socket> with verification for <host>" variant.
 	MCExpression *verifyhostname;
+    MCExpression *fromaddress;
```
I suggest using an `MCAutoPointer<MCExpression>` here.  Then it will be automatically initialised and cleaned up, and you can omit the changes to the `MCOpen` constructor and destructor.

peter-b commented on engine/src/socket.h:
```diff
@@ -142,8 +142,9 @@ class MCSocket
 	MCSocketHandle fd;	
 	// MM-2014-06-13: [[ Bug 12567 ]] Added support for specifying an end host name to verify against.
 	MCNameRef endhostname;
+    MCNameRef from;
```
I suggest using an `MCNewAutoNameRef` here.

peter-b reviewed (commented):
Nice! Just a couple of suggested changes to using managed pointer types in structures.

livecodemichael committed:
[[ SocketLocalPort ]] Fixed errors in dictionary entry.

livecodemichael committed:
[[ SocketLocalPort ]] Added tests.

livecodemichael committed:
[[ SocketLocalPort ]] Updated to use managed pointer types where
necessary.

livecodemichael commented:
@peter-b Thanks for the suggestions. That should be PR updated to take them into account.

peter-b reviewed (commented):
@livecodemichael Could you please add a test that this works correctly when you specify `from` clause with an address part but no port part?  I think in that case it should work, but should bind to any available port.  It would probably be a good idea to add some tests for `from` clauses with both address and port parts, as well, please!

livecodemichael committed:
[[ SocketLocalPort ]] Added tests for checking specifying a local
address works with open socket from.

livecodemichael commented:
@peter-b Done!

livecodemichael committed:
[[ SocketLocalPort ]] Added socket tests that use a local server.

peter-b commented:
@livecodemichael Thank you very much for adding the extra tests!

@livecode-vulcan review ok c380b0b

livecode-vulcan commented:
💙 review by @peter-b ok c380b0b3d63b73ce9c99b77a8bd660e81087e27e

livecode-vulcan commented:
😞 test failure c380b0b3d63b73ce9c99b77a8bd660e81087e27e

* try-community-linux-x86: [failure](https://vulcan.livecode.com/builders/try-community-linux-x86/builds/976)
* try-community-mac: [pending](https://vulcan.livecode.com/builders/try-community-mac/builds/982)
* try-community-linux-x86_64: [failure](https://vulcan.livecode.com/builders/try-community-linux-x86_64/builds/974)
* try-community-emscripten: [pending](https://vulcan.livecode.com/builders/try-community-emscripten/builds/975)
* try-community-ios: [pending](https://vulcan.livecode.com/builders/try-community-ios/builds/982)
* try-community-win-x86: [pending](https://vulcan.livecode.com/builders/try-community-win-x86/builds/996)
* try-community-android-armv6: [failure](https://vulcan.livecode.com/builders/try-community-android-armv6/builds/973)

livecodemichael committed:
[[ SocketLocalPort ]] Fix compitation errors.

SO_REUSEPORT is not defined on all platforms so use SO_REUSEADDR.
In our case, SO_REUSEADDR is probabaly the better option to use.

Fixed missing arg to MCSocket constructor in win only code path.

peter-b commented:
@livecode-vulcan review ok c5dfe71

livecode-vulcan commented:
💙 review by @peter-b ok c5dfe7130d8ee4e8308aa5dba8435b2597c28954

livecode-vulcan commented:
😎 test success c5dfe7130d8ee4e8308aa5dba8435b2597c28954

* try-community-linux-x86: [success](https://vulcan.livecode.com/builders/try-community-linux-x86/builds/977)
* try-community-mac: [success](https://vulcan.livecode.com/builders/try-community-mac/builds/983)
* try-community-linux-x86_64: [success](https://vulcan.livecode.com/builders/try-community-linux-x86_64/builds/975)
* try-community-emscripten: [success](https://vulcan.livecode.com/builders/try-community-emscripten/builds/976)
* try-community-ios: [success](https://vulcan.livecode.com/builders/try-community-ios/builds/983)
* try-community-win-x86: [success](https://vulcan.livecode.com/builders/try-community-win-x86/builds/997)
* try-community-android-armv6: [success](https://vulcan.livecode.com/builders/try-community-android-armv6/builds/974)
